src/lib/io/sol-iio.c

```diff
@@ -700,12 +733,31 @@ sol_iio_open(int device_id, const struct sol_iio_config *config)
         return NULL;
     }
 #endif
+    SOL_VECTOR_FOREACH_IDX (&iio_opened_devices, entry, i) {
+        if (device_id == entry->device->device_id) {
+            if ((config->buffer_size == entry->config->buffer_size)
+                && (config->sampling_frequency == entry->config->sampling_frequency)) {
+                entry->refcount++;
+                device = entry->device;
+                reader_cb_data_elem = sol_vector_append(&device->reader_cb_list);
+                SOL_NULL_CHECK(reader_cb_data_elem, NULL);
```

Comment:
If this check fails, you'd need to decrease `refcount`. Increase `refcount` just before `return device` to keep things simpler.